community/lucene-index/src/main/java/org/neo4j/kernel/api/impl/schema/verification/CompositeDuplicateCheckingCollector.java

```diff
@@ -0,0 +1,140 @@
+ * along with this program.  If not, see <http://www.gnu.org/licenses/>.
+ */
+package org.neo4j.kernel.api.impl.schema.verification;
+
+import org.apache.lucene.document.Document;
+import org.apache.lucene.index.LeafReader;
+import org.apache.lucene.index.LeafReaderContext;
+import org.apache.lucene.search.SimpleCollector;
+
+import java.io.IOException;
+import java.util.Arrays;
+
+import org.neo4j.kernel.api.StatementConstants;
+import org.neo4j.kernel.api.exceptions.KernelException;
+import org.neo4j.kernel.api.exceptions.index.IndexEntryConflictException;
+import org.neo4j.kernel.api.impl.schema.LuceneDocumentStructure;
+import org.neo4j.kernel.api.index.PropertyAccessor;
+import org.neo4j.kernel.api.properties.Property;
+import org.neo4j.kernel.api.schema_new.OrderedPropertyValues;
+
+public class CompositeDuplicateCheckingCollector extends DuplicateCheckingCollector
+{
+    private final int[] propertyKeyIds;
+    private CompositeEntrySet actualValues;
+
+    public CompositeDuplicateCheckingCollector( PropertyAccessor accessor, int[] propertyKeyIds )
+    {
+        super(accessor, -1);
+        this.propertyKeyIds = propertyKeyIds;
+        actualValues = new CompositeEntrySet();
+    }
+
+    @Override
+    protected void doCollect( int doc ) throws IOException, KernelException, IndexEntryConflictException
+    {
+        Document document = reader.document( doc );
+        long nodeId = LuceneDocumentStructure.getNodeId( document );
+        Property[] properties = new Property[propertyKeyIds.length];
+        Object[] values = new Object[propertyKeyIds.length];
+        for ( int i = 0; i < properties.length; i++ )
+        {
+            properties[i] = accessor.getProperty( nodeId, propertyKeyIds[i] );
+            values[i] = properties[i].value();
+        }
+
+        // We either have to find the first conflicting entry set element,
+        // or append one for the property we just fetched:
+        CompositeEntrySet current = actualValues;
+        scan:
+        do
+        {
+            for ( int i = 0; i < CompositeEntrySet.INCREMENT; i++ )
+            {
+                Object[] currentValues = current.values[i];
+
+                if ( current.nodeId[i] == StatementConstants.NO_SUCH_NODE )
+                {
+                    current.values[i] = values;
+                    current.nodeId[i] = nodeId;
+                    if ( i == CompositeEntrySet.INCREMENT - 1 )
+                    {
+                        current.next = new CompositeEntrySet();
+                    }
+                    break scan;
+                }
+                else if ( propertyValuesEqual( properties, currentValues ) )
+                {
+                    throw new IndexEntryConflictException( current.nodeId[i], nodeId,
+                            OrderedPropertyValues.ofUndefined( currentValues ) );
+                }
+            }
+            current = current.next;
+        }
+        while ( current != null );
+    }
```

Comment:
I wonder if there's more duplication between this class and its parent, that could be removed here.